### PR TITLE
Allow handling drawable input event externally

### PR DESCRIPTION
**Pull request by luaneko**
Adds a public `OnInput` event that is raised by `Handle(UIEvent)` and can be registered from outside. This is useful when extending the input handling logic of some drawable without having to derive it.

### CODE REVIEW

File: osu.Framework/Graphics/Drawable.cs [hunk 1]
```diff
@@ -2195,15 +2201,26 @@ public bool TriggerEvent(UIEvent e)
...
+        private bool requestsNonPositionalCached;
+        private bool requestsPositionalCached;
+
         /// <summary>
         /// Whether this drawable should receive non-positional input. This does not mean that the drawable will immediately handle the received input, but that it may handle it at some point.
         /// </summary>
-        internal bool RequestsNonPositionalInput { get; private set; }
+        internal bool RequestsNonPositionalInput
+        {
+            get => requestsNonPositionalCached || OnInput != null;
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Drawable.cs]: What happens if the delegate is set after it’s constructed, like say, inside BDL `load()`?

**Comment by peppy** [comment]
Not sure about this. We intentionally limit publicly exposing things like this to (hopefully) lead to more well-thought design. Can you give a use case where you would be seeing this as useful?

**Comment by phosphene47** [comment]
In my specific case, I'm writing some drawable wrappers for my [ofreact](https://github.com/phosphene47/ofreact) project where consumers need to define input handlers declaratively rather than by deriving drawables.

This is because composition is preferred over inheritance in declarative programming. There is a set of built-in drawable-wrapping components like `Text`, `Box`, `Container`, etc. that form the building blocks of a more complex component; they are akin to sealed classes. And they need to receive input handling delegates from the rendering function of a more complex component, to *set* on the drawable.

Example pseudocode: This is a normal osu!framework drawable that uses inheritance to add input logic.

```csharp
class mytext : SpriteText {
  ctor() { Text = "my text"; }

  override OnMouseDown() => Text = "another text";
}
```

This would be equivalent to:

```csharp
class mytext : declarative Drawable {
  override Render() {
    var (text, setText) = UseState("my text");

    return new SpriteTextWrapper(
      text: text,
      onMouseDown: () => setText("another text"));
  }
}
```

`SpriteTextWrapper` class represents a `SpriteText` drawable in the declarative component tree. Notice that it is taking a delegate via constructor parameter: this delegate would be assigned to the `OnInput` event of `SpriteText`.

**Comment by smoogipoo** [comment]
Since you're creating Drawables dynamically, maybe a better choice is to dynamically override `Handle()` to invoke your event. Going forward you could use this to apply hooks other than just to `Handle()`.

Something like https://docs.microsoft.com/en-us/dotnet/api/system.reflection.emit.typebuilder.definemethodoverride?view=netstandard-2.1 may be what I'm imagining.

**Comment by smoogipoo** [comment]
Let's come back to this one when implementing OML (again).

**Pull request closed without merging**